### PR TITLE
Feature - add pr labels to integration pipelines

### PR DESCRIPTION
**Pull request by mariana-dias-alves**
- Add pipeline for fetching PR labels
- Add pipeline to close PRs
- Adapt integration pipeline to deploy on PR to CI. Retro-compatibility tested here with release 0.0.0-1: https://github.com/MOV-AI/project-tugbot2/actions/runs/10735968725
     - Generated https://registry.harbor.es.mov.ai/harbor/projects/3/repositories/spawner-tugbot2-noetic/artifacts-tab/artifacts/sha256:c8f0beaefde6462b3be18ad461244850bb5548c85e5b1647198d4cfb64f0396e?sbomDigest=
- Add pipeline to comment on PR at end of deploy to CI pipeline, such as:
image
 

-----------------

- [x] Make sure you are opening from a **topic/feature/bugfix branch**
- [x] Ensure that the PR title represents the desired changes
- [x] Ensure that the PR description detail the desired changes
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

[^note]:
    Put an `x` into the [ ] to show you have filled the information.
    The template comes from https://github.com/MOV-AI/.github/blob/master/.github/pull_request_template.md
    You can override it by creating .github/pull_request_template.md  in your own repository

### CODE REVIEW

File: .github/workflows/integration-build-product.yml [hunk 1]
```diff
@@ -56,6 +56,13 @@ on:
         required: false
         type: boolean
         default: false
+      pr_id:
```
Comment (duartecoelhomovai) [inline .github/workflows/integration-build-product.yml:59]: test the inputs not mentioned and you'll have my okay

**Review by duartecoelhomovai — CHANGES_REQUESTED** [review]

File: .github/workflows/clean-closed-prs.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+jobs:
+  handle-pr-merge:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        distro: ${{ fromJSON(inputs.ros_distro) }}
+        registry: ${{ fromJSON(inputs.registries) }}
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v3
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/clean-closed-prs.yml]:
```suggestion
        uses: actions/checkout@v4
```

File: .github/workflows/clean-closed-prs.yml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        run: |
+          sudo apt install skopeo -y
+
+          PR_NUMBER=${{ github.event.pull_request.number }}
+          TAG="pr${PR_NUMBER}"
+          echo Deleting images with tags related to $TAG
+
+          echo "Cleaning images from ${{ matrix.registry }}"
+
+          namespaces=(prod qa ci devops)
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/clean-closed-prs.yml]: I would rather not act on prod and qa by default

File: .github/workflows/fetch-pr-info.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            LABELS=$(echo "${{ toJson(github.event.pull_request.labels.*.name) }}" | tr -d '[] ' | sed '/^$/d' | tr '\n' ','  | sed 's/^/[/' | sed 's/,$/]/')
+            # Set the output values
+            echo "pr_number=${PR_NUMBER}" | tee >> $GITHUB_OUTPUT
+            echo "pr_labels='[${LABELS}]'" | tee >> $GITHUB_OUTPUT
+          else
+            echo "This is neither a push event nor a pull_request event."
+            echo "pr_number=" >> $GITHUB_OUTPUT
+            echo "pr_labels='[]'"  >> $GITHUB_OUTPUT
+          fi
+
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/fetch-pr-info.yml]: too many empty lines

File: .github/workflows/integration-build-product.yml [hunk 5]
```diff
@@ -230,7 +252,7 @@ jobs:
           integration-pipeline generate_meta_simulator_artifacts \
                 --manifest_platform_base_key product_dependencies \
                 --product_name ${{ inputs.product_name }} \
-                --branch ${BRANCH_NAME} \
+                --branch $BRANCH_NAME \
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/integration-build-product.yml]: why this change ? (out of curiosity)

File: .github/workflows/integration-build-product.yml [hunk 6]
```diff
@@ -252,12 +274,25 @@ jobs:
           rm -rf simulator_artifacts ci_artifacts
           mkdir platform_configs
           source ci_scripts/bin/activate
-          integration-pipeline raise
+          # Check if a PR triggered the workflow
+          pip install bump2version
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/integration-build-product.yml]: should bump2version already be in our container image ?

File: .github/workflows/integration-build-product.yml [hunk 7]
```diff
@@ -529,7 +584,21 @@ jobs:
           docker pull "${{ env.MID_REGISTRY }}/ci/${{ inputs.product_name }}-${{ matrix.distro }}:${{ needs.Build-Spawner.outputs.raised_version }}"
           docker tag "${{ env.MID_REGISTRY }}/ci/${{ inputs.product_name }}-${{ matrix.distro }}:${{ needs.Build-Spawner.outputs.raised_version }}" "${{ env.REGISTRY }}/qa/${{ inputs.product_name }}-${{ matrix.distro }}:${{ needs.Build-Spawner.outputs.raised_version }}"
           docker tag "${{ env.MID_REGISTRY }}/ci/${{ inputs.product_name }}-${{ matrix.distro }}:${{ needs.Build-Spawner.outputs.raised_version }}" "${{ env.PUSH_REGISTRY }}/qa/${{ inputs.product_name }}-${{ matrix.distro }}:${{ needs.Build-Spawner.outputs.raised_version }}"
-
+      
+      - name: Update robot jsons to use qa project
+        if: ${{ github.event_name == 'pull_request' }}
+        shell: bash
+        run: |
+          # iterate over .json files and update spawner.image to use ci image
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/integration-build-product.yml]: repeated code here...
can't we prepare, store and reuse the jsons instead of creating them twice ?

Comment (AlexFernandes-MOVAI) [reply .github/workflows/integration-build-product.yml]: maybe in `Validate-boostrap-configs` step ?

**Review by AlexFernandes-MOVAI — CHANGES_REQUESTED** [review]
I'm sure it works but I'm seeing some small and easy improvements to add :-s

File: .github/workflows/integration-build-product.yml (hunk 5 above)
Comment (mariana-dias-alves) [reply .github/workflows/integration-build-product.yml]: no reason, reverting

File: .github/workflows/clean-closed-prs.yml (hunk 3 above)
Comment (mariana-dias-alves) [reply .github/workflows/clean-closed-prs.yml]: done

File: .github/workflows/integration-build-product.yml (hunk 6 above)
Comment (mariana-dias-alves) [reply .github/workflows/integration-build-product.yml]: it current is not, but maybe could be added

**Review by duartecoelhomovai — APPROVED** [review]

File: .github/workflows/integration-build-product.yml [hunk 8]
```diff
@@ -379,11 +379,13 @@ jobs:
         run: |
           . .ci-venv/bin/activate
           # iterate over .json files and update spawner.image to use ci image
+          mkdir qa
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/integration-build-product.yml]: what does qa means here ? does not seem to be used after you copy the jsons there .... maybe I miss something ...

**Review by AlexFernandes-MOVAI — COMMENTED** [review]
clarify qa dir please

File: .github/workflows/integration-build-product.yml (hunk 6 above)
Comment (AlexFernandes-MOVAI) [reply .github/workflows/integration-build-product.yml]: let's add it then instead of reinstalling it 10 times a day

File: .github/workflows/integration-build-product.yml (hunk 8 above)
Comment (mariana-dias-alves) [reply .github/workflows/integration-build-product.yml]: removing it

File: .github/workflows/integration-build-product.yml [hunk 9]
```diff
@@ -877,6 +939,7 @@ jobs:
           echo "msg_error=${MESSAGE_ERR}\n  Details: ${{ steps.job_info.outputs.job_url }}" >> $GITHUB_OUTPUT
 
       - name: Slack message success
+        if: ${{ github.event_name != 'pull_request' }}
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/integration-build-product.yml:951]: please post a comment on the PR which indicates at least:
-  artifacts were generated and available at url: https://github.com/MOV-AI/project-tugbot2/actions/runs/10740271564 for example
- artifacts will expire in x days or after x operation
- how robot.json can be used to test locally

see: https://github.com/peter-evans/create-or-update-comment

**Review by AlexFernandes-MOVAI — CHANGES_REQUESTED** [review]
- adding or updating a comment to communicate artifacts were generated would simplify navigation and reviews
- removing installation of bump2version will speed up a bit the PR checks

File: .github/workflows/integration-build-product.yml (hunk 9 above)
Comment (mariana-dias-alves) [reply .github/workflows/integration-build-product.yml:951]: done

File: .github/workflows/comment-on-pr.yml [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        runs-on: ubuntu-latest
+        steps:
+          - name: Get Artifact URL & PR Info
+            env:
+              GITHUB_TOKEN: ${{ secrets.gh_token }}
+              OWNER: ${{ github.repository_owner }}
+              REPO: ${{ github.event.repository.name }}
+              WORKFLOW_RUN_EVENT_OBJ: ${{ toJSON(github.event.workflow_run) }}
+            run: |
+              PREVIOUS_JOB_ID=$(jq -r '.id' <<< "$WORKFLOW_RUN_EVENT_OBJ")
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/comment-on-pr.yml:20]: as the name is previous, I always wonder what if this is the first iteration and there is no previous job ?

File: .github/workflows/comment-on-pr.yml [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+              echo "Previous Job ID: $PREVIOUS_JOB_ID"
+              echo "PREVIOUS_JOB_ID=$PREVIOUS_JOB_ID" >> "$GITHUB_ENV"
+    
+              PR_NUMBER=$(jq -r '.pull_requests[0].number' \
+                  <<< "$WORKFLOW_RUN_EVENT_OBJ")
```
Comment (AlexFernandes-MOVAI) [inline .github/workflows/comment-on-pr.yml:25]: wahou, first time I see so many usages of `<<<` , is it really needed ?

from `man bash`:

```
Here Strings
       A variant of here documents, the format is:

              [n]<<<word

       The word undergoes tilde expansion, parameter and variable expansion, command substitution, arithmetic expansion, and quote removal.  Pathname expansion and word splitting are not performed.  The result is supplied as a
       single string, with a newline appended, to the command on its standard input (or file descriptor n if n is specified).
```

File: .github/workflows/comment-on-pr.yml (hunk 10 above)
Comment (mariana-dias-alves) [reply .github/workflows/comment-on-pr.yml:20]: this workflow is always triggered by the completion of a workflow of type "Deploy to CI", which is the workflow from which I am getting the id, and that is why the PREVIOUS prefix is used. So there will always be a previous workflow

File: .github/workflows/comment-on-pr.yml (hunk 11 above)
Comment (mariana-dias-alves) [reply .github/workflows/comment-on-pr.yml:25]: I used this snippet of code https://github.com/orgs/community/discussions/51403#discussioncomment-5515349 which used `<<<`. I am unsure if it is absolutely necessary

**Review by AlexFernandes-MOVAI — APPROVED** [review]
